Extend timer.Timer to include startup method

keithhamilton opened the pull request:
In working with the `rospy.timer.Timer` class, it feels to me like it could be improved with two subtle changes:
- A kwarg in its constructor that allows its invoker to either instantiate it in a running or shutdown state
- A companion method to `shutdown` that starts running the timer instead of turning it off.

This PR adds a kwarg to the `timer.Timer` constructor, `start_running` that defaults to False. In the `__init__` method, `self._shutdown` is set to `not start_running` so that if you wish to start the Timer in a running state, `start_running=True` and `self._shutdown = False`.

This PR also adds a `startup` method as a companion to `shutdown`, adding the ability to retain one instance of a `timer.Timer` object with the ability to toggle its running state. 

I have found this great for establishing a timer that controls a certain behavior with our robot, that may start and stop depending on repeating conditions.

Thanks for checking this out!
-- Keith

Commit:
Timer defaults to shutdown; added start toggle

Commit:
renamed start -> startup

ros-pull-request-builder commented:
Can one of the admins verify this patch?

Commit:
Changed start_running default to True

This way backwards compatibility is not affected.

Commit:
added additional __period member

using the second member the initial period can be retained and the timer can be stopped/started as many times as needed.

dirk-thomas commented on clients/rospy/src/rospy/timer.py:
```diff
@@ -48,7 +48,7 @@ class Rate(object):
     Convenience class for sleeping in a loop at a specified rate
     """
-    
+
```
Please avoid any white space change.

dirk-thomas commented on clients/rospy/src/rospy/timer.py:
```diff
@@ -190,12 +190,15 @@ def __init__(self, period, callback, oneshot=False):
         @param oneshot: if True, fire only once, otherwise fire continuously until shutdown is called [default: False]
         @type  oneshot: bool
+        @param start: if True, begin running the `run` method immediately, otherwise wait until self.start() is called [default: False]
+        @type  start: bool
```
The name in the documentation does not match the new argument.

dirk-thomas commented on clients/rospy/src/rospy/timer.py:
```diff
@@ -190,12 +190,15 @@ def __init__(self, period, callback, oneshot=False):
         """
         threading.Thread.__init__(self)
-        self._period   = period
+        self.__period = period
```
What is this needed for? `self._period` is never altered anywhere.

dirk-thomas commented on clients/rospy/src/rospy/timer.py:
```diff
@@ -204,7 +207,14 @@ def shutdown(self):
-        
+        self._period = self.__period
+
+    def startup(self):
```
It is not clear to me how this is supposed to be used by the user. If the timer has been shut down before and the thread has terminated already calling this function has no effect on its own. The user also has to call `start()` again.

If the thread is still running calling this method _undos_ a previous `shutdown` call which seems to be unintuitive.

keithhamilton replied:
The point is that "shutting down" a threading.Thread object doesn't actually shut it down, right? It just flips a bit that causes the `run` method to ignore executing its main logic block. (line 222)

The thread is technically still running until its out of scope or is assigned a NoneType value. 

My use case was having a Timer object that I could start, shutdown, retain, and restart as a condition changed.

keithhamilton replied:
Also, calling `start()` is not necessary. In this case you have a Timer object whose `_shutdown` property is True (`run` is not executing its main logic block), yet the thread still exists. Calling `startup()` flips the bit so that the `run` method begins paying attention to its main logic block again.

dirk-thomas replied:
Have you tried to use the timer in this way? I don't think this is the behavior of a Python `Thread` instance. The `run()` is invoked **once** when the thread is being started. As soon as the run function returns the thread is done.

When the `_shutdown` flag is being set the while loop will be stopped and the run function returns and therefore the thread is finished.

keithhamilton replied:
What's the protocol for suddenly realizing I had been misinterpreting something rather key to the functionality of a Thread object, and now feel like an idiot?

I was mistaking the run method for **not** exiting when `_shutdown` was flipped to `True`, hence the idea of making it possible to continue its existence after telling it to stop. I suppose it would be possible to have an implementation that did that, but as-is, the code in this PR does not make that functionally possible..

Sorry bout that! Feel free to close this with a vengeance.

Commit:
Update start_running param in docstring

It previously was erroneously named `start`, whereas the correct name, in the method signature is `start_running`

Commit:
Remove whitespace change

dirk-thomas commented:
Since a different approach is necessary to achieve the goal of the feature request I will close this PR for now.
